Objects/descrobject.c

```diff
@@ -1790,3 +1790,152 @@ PyTypeObject PyProperty_Type = {
+}
+
+static int
+ga_traverse(PyObject *self, visitproc visit, void *arg)
+{
+    gaobject *alias = (gaobject *)self;
+    Py_VISIT(alias->origin);
+    Py_VISIT(alias->parameters);
+    return 0;
+}
+
+// TODO: nicely format origin and each parameter
+static PyObject *
+ga_repr(PyObject *self)
+{
+    gaobject *alias = (gaobject *)self;
+    return PyUnicode_FromFormat("%R[%R]", alias->origin, alias->parameters);
+}
+
+static PyObject *
+ga_call(PyObject *self, PyObject *args, PyObject *kwds)
+{
+    gaobject *alias = (gaobject *)self;
+    return PyObject_Call(alias->origin, args, kwds);
+}
+
+static const char* attr_exceptions[] = {
+    "__origin__",
+    "__parameters__",
+    "__mro_entries__",
+    NULL,
+};
+
+static PyObject *
+ga_getattro(PyObject *self, PyObject *name)
+{
+    gaobject *alias = (gaobject *)self;
+    if (alias->origin != NULL && PyUnicode_Check(name)) {
+        for (const char **p = attr_exceptions; ; p++) {
+            if (*p == NULL) {
+                return PyObject_GetAttr(alias->origin, name);
+            }
+            if (PyUnicode_CompareWithASCIIString(name, *p) == 0) {
+                break;
+            }
+        }
+    }
+    return PyObject_GenericGetAttr(self, name);
+}
+
+static PyObject *
+ga_mro_entries(PyObject *self, PyObject *args)
+{
+    gaobject *alias = (gaobject *)self;
+    return Py_BuildValue("(O)", alias->origin);
+}
+
+static PyMethodDef ga_methods[] = {
+    {"__mro_entries__", ga_mro_entries, METH_O},
+    {0}
+};
+
+static PyMemberDef ga_members[] = {
+    {"__origin__", T_OBJECT, offsetof(gaobject, origin), READONLY},
+    {"__parameters__", T_OBJECT, offsetof(gaobject, parameters), READONLY},
+    {0}
+};
+
+static PyObject *
+ga_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+    if (kwds != NULL) {
+        return PyErr_Format(PyExc_TypeError, "GenericAlias does not support keyword arguments");
+    }
+    if (!PyTuple_Check(args) || PyTuple_Size(args) != 2) {
+        return PyErr_Format(PyExc_TypeError, "GenericAlias expects 2 positional arguments");
+    }
+    PyObject *origin = PyTuple_GET_ITEM(args, 0);
+    PyObject *parameters = PyTuple_GET_ITEM(args, 1);
+    return Py_GenericAlias(origin, parameters);
+}
+
+// TODO:
+// - argument clinic?
+// - __doc__?
+// - cache?
+PyTypeObject Py_GenericAliasType = {
+    PyVarObject_HEAD_INIT(&PyType_Type, 0)
+    .tp_name = "GenericAlias",
+    .tp_basicsize = sizeof(gaobject),
+    .tp_dealloc = ga_dealloc,
+    .tp_repr = ga_repr,
+    .tp_call = ga_call,
+    .tp_getattro = ga_getattro,
+    .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC,
+    .tp_traverse = ga_traverse,
+    .tp_methods = ga_methods,
+    .tp_members = ga_members,
+    .tp_alloc = PyType_GenericAlloc,
+    .tp_new = ga_new,
+    .tp_free = PyObject_GC_Del,
+};
+
+PyObject *
+Py_GenericAlias(PyObject *origin, PyObject *parameters)
+{
+    PyObject *wrapper = NULL;
+    if (!PyTuple_Check(parameters)) {
+        wrapper = PyTuple_New(1);
```

Comment:
```c
if (!PyTuple_Check(parameters)) {
    parameters = PyTuple_Pack(1, parameters);
    if (parameters == NULL) {
        return NULL;
    }
}
else {
    Py_INCREF(parameters);
}
```
Also add `Py_DECREF(parameters)` after `if (alias == NULL)` and remove `Py_INCREF(parameters)` below.